service/geopmdpy/system_files.py

```diff
@@ -728,6 +735,61 @@ def remove_batch_server(self, client_pid):
             sys.stderr.write(f'Warning: {write_fifo_path} file was left over, deleting it now.\n')
             os.unlink(write_fifo_path)
 
+    def start_profile(self, client_pid, profile_name):
+        profile_name = str(profile_name)
+        self.check_client_active(client_pid, 'start_profile')
+        uid, gid = self._pid_info(client_pid)
+        self._sessions[client_pid]['client_uid'] = int(uid)
+        self._sessions[client_pid]['client_gid'] = int(gid)
+        if len(self._profiles) == 0:
+            size = 64 * os.cpu_count()
+            shmem.create_prof('status', size, client_pid, uid, gid)
+        if profile_name in self._profiles:
+            self._profiles[profile_name].add(client_pid)
+        else:
+            self._profiles[profile_name] = {client_pid}
+        self._sessions[client_pid]['profile_name'] = profile_name
+        size = 57384
+        shmem.create_prof('record-log', size, client_pid, uid, gid)
+        self._update_session_file(client_pid)
```

Comment:
There is a scenario that could cause problems with this code. Given how the client is going to call into these APIs, it should never happen, but it's still worth protecting against in case of a bug in the client or a malicious actor calling into the DBus APIs.

The scenario happens as a result of:
- A client can call `PlatformStartProfile` multiple times
- The code tracks the profile a client belongs to in two data structures: a map of profile names to PIDs (as a profile can have multiple PIDs), and the session data map (where a PID's profile name is stored)

The scenario is: a client calls `PlatformStartProfile` twice with two different profile names. The way the code currently is, it will add the client's PID to two different profiles in the profiles map, and would only track the last profile name with which it was called in the session data map. So when this client stops profiling, its profile will be removed from the session data map, but it will be removed only from one of the profiles in the profiles map.

I think the `shmem.create_prof()` call might fail in this scenario the second time its called and might prevent this scenario from happening, but I think it would be better to guard against it more explicitly.